### PR TITLE
Write more on Animation and related docs

### PR DESCRIPTION
**Pull request by gnprice**
A few days ago I started reading up on how animations work, in
preparation for starting to use them in a more complex way than
I'd done before.  I found it a bit difficult to get my head around;
in particular the many different classes involved, how they relate
to each other, and how to fit them together.

So once I had worked that out, I sat down to express it in the form
of documentation.

The largest change here is an expansion of the docs on [Animation]
itself, including a new section "Using animations" with several
paragraphs laying out how one typically fits together
AnimationController, TickerProvider, CurvedAnimation and/or Tween,
and AnimatedWidget subclasses.  [Animation] also gets an expanded
"See also" list, a revised conceptual intro, and a new summary line.

There are also revisions on [TickerProvider], [AnimatedController],
and elsewhere; some new exposition, some revisions for clarity, and
various small fixes.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/animation/animation.dart [hunk 1]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
-/// To create a new animation that you can run forward and backward, consider
-/// using [AnimationController].
+/// An animation may move forward or backward on its own as time passes
+/// (like the opacity of a button that fades over a fixed duration
+/// once the user touches it),
+/// or it may be driven by the user
+/// (like the position of a slider that the user can drag back and forth),
+/// or it may do both
+/// (like a switch that snaps into place when releaseed,
+/// or a [Dismissible] that responds to drag and fling gestures, etc.).
```
Comment (loic-sharma) [inline packages/flutter/lib/src/animation/animation.dart:86]:
```suggestion
/// An animation can move forward or backward on its own as time passes
/// (like the opacity of a button that fades over a fixed duration
/// once the user touches it),
/// or it can be driven by the user
/// (like the position of a slider that the user can drag back and forth),
/// or it can do both
/// (like a switch that snaps into place when releaseed,
/// or a [Dismissible] that responds to drag and fling gestures, etc.).
```

File: packages/flutter/lib/src/animation/animation.dart [hunk 2]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
+///
+/// ## Using animations
+///
+/// For simple animation effects, consider using one of the
+/// [ImplicitlyAnimatedWidget] subclasses,
+/// like [AnimatedScale], [AnimatedOpacity], and many others.
+/// When an [ImplicitlyAnimatedWidget] suffices, there is
+/// no need to work with [Animation], [AnimationController], or [Tween],
+/// or even to have a [State] to manage the animation (though all of those are
+/// involved in the implementation of the [ImplicitlyAnimatedWidget] itself).
```
Comment (loic-sharma) [inline packages/flutter/lib/src/animation/animation.dart:101]: For folks that want simple animation effects, this might be too much context and might be overwhelming. Perhaps we should simplify this section?

```suggestion
/// Widgets of this type animate when their inputs change.
```

File: packages/flutter/lib/src/animation/animation.dart [hunk 3]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
+///
+/// Otherwise, typically an animation originates with an [AnimationController]
+/// (which is itself an [Animation<double>])
+/// created by a [State] that implements [TickerProvider],
+/// and then further animations may be derived from it using
+/// e.g. [Tween] or [CurvedAnimation].
```
Comment (loic-sharma) [inline packages/flutter/lib/src/animation/animation.dart:107]: This sentence introduces quite a few concepts. Maybe consider splitting it in two:

```suggestion
/// Otherwise, typically an animation originates with an [AnimationController]
/// (which is itself an [Animation<double>])
/// created by a [State] that implements [TickerProvider].
/// Further animations can be derived from it using
/// e.g. [Tween] or [CurvedAnimation].
```

File: packages/flutter/lib/src/animation/animation.dart [hunk 4]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
+/// or their values might be used directly.
+///
+/// For example, the [AnimationController] may represent
+/// the abstract progress of the animation from 0.0 to 1.0;
+/// then a [CurvedAnimation] might apply an easing curve;
+/// and a [SizeTween] and [ColorTween] might each be applied to that
+/// to produce an [Animation<Size>] and an [Animation<Color>] that control
+/// a widget shrinking and changing color as the animation proceeds.
+///
+/// Because the [Animation] keeps the same identity as the animation proceeds,
```
Comment (loic-sharma) [inline packages/flutter/lib/src/animation/animation.dart:119]: This is an excellent tip. Do you think it deserves its own section?

```suggestion
/// ## Performance
///
/// Because the [Animation] keeps the same identity as the animation proceeds,
```

File: packages/flutter/lib/src/animation/animation.dart [hunk 5]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
+/// it provides a convenient way for a [StatefulWidget] that orchestrates
+/// a complex animation to communicate the animation's progress to its
+/// various child widgets.  Consider having higher-level widgets in the tree
+/// pass lower-level widgets the [Animation] itself, rather than its value,
+/// in order to avoid rebuilding the higher-level widgets on each frame
+/// even while the animation is active.
+/// If the leaf widgets also ignore [value] and pass the whole
+/// [Animation] object to a render object (like [FadeTransition] does),
+/// they too may be able to avoid rebuild and even relayout, so that the
```
Comment (loic-sharma) [inline packages/flutter/lib/src/animation/animation.dart:128]:
```suggestion
/// they too might be able to avoid rebuild and even relayout, so that the
```

File: packages/flutter/lib/src/animation/animation.dart [hunk 6]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
+/// (like the opacity of a button that fades over a fixed duration
+/// once the user touches it),
+/// or it may be driven by the user
+/// (like the position of a slider that the user can drag back and forth),
+/// or it may do both
+/// (like a switch that snaps into place when releaseed,
+/// or a [Dismissible] that responds to drag and fling gestures, etc.).
+/// The behavior is normally controlled by method calls on
+/// some underlying [AnimationController],
+/// for example by a [State] that created the [AnimationController].
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation.dart:89]: Not a huge deal, but I did have to re-read this a couple of times, since at first glance it sounded like a `State` is an example of an `AnimationController`.

Perhaps it could just be deleted, since it's also explained [further down](https://github.com/flutter/flutter/blob/c15bf8ead6d07dfa1755fba41c1c8ed0c81a50f4/packages/flutter/lib/src/animation/animation.dart#L103-L107).

```suggestion
```

File: packages/flutter/lib/src/animation/animation.dart [hunk 7]
```diff
@@ -60,33 +60,93 @@ typedef AnimationStatusListener = void Function(AnimationStatus status);
...
+/// to produce an [Animation<Size>] and an [Animation<Color>] that control
+/// a widget shrinking and changing color as the animation proceeds.
+///
+/// Because the [Animation] keeps the same identity as the animation proceeds,
+/// it provides a convenient way for a [StatefulWidget] that orchestrates
+/// a complex animation to communicate the animation's progress to its
+/// various child widgets.  Consider having higher-level widgets in the tree
+/// pass lower-level widgets the [Animation] itself, rather than its value,
+/// in order to avoid rebuilding the higher-level widgets on each frame
+/// even while the animation is active.
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation.dart:125]:
```suggestion
/// while the animation is active.
```

File: packages/flutter/lib/src/animation/animation_controller.dart [hunk 8]
```diff
@@ -81,34 +81,29 @@ enum AnimationBehavior {
...
-///
-/// The widget test framework [WidgetTester] object can be used as a ticker
-/// provider in the context of tests. In other contexts, you will have to either
-/// pass a [TickerProvider] from a higher level (e.g. indirectly from a [State]
-/// that mixes in [TickerProviderStateMixin]), or create a custom
-/// [TickerProvider] subclass.
+/// For advice on obtaining a ticker provider, see [TickerProvider].
+/// Typically the relevant [State] serves as the ticker provider,
+/// after applying a suitable mixin to cause the [State] subclass to
+/// implement [TickerProvider].
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:106]: I much prefer the more concise version, thank you 🙂

That being said, I like how the original explicitly mentions `TickerProviderStateMixin` and `SingleTickerProviderStateMixin`, since that's the most important thing to know when setting up an `AnimationController`.

File: packages/flutter/lib/src/animation/tween.dart [hunk 9]
```diff
@@ -82,9 +82,10 @@ abstract class Animatable<T> {
   }
 
   /// Returns a new [Animatable] whose value is determined by first evaluating
-  /// the given parent and then evaluating this object.
+  /// the given parent and then evaluating this object on the result.
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/tween.dart:85]:
```suggestion
  /// the given parent and then evaluating this object based on the result.
```

File: packages/flutter/lib/src/scheduler/ticker.dart [hunk 10]
```diff
@@ -14,23 +14,28 @@ export 'package:flutter/foundation.dart' show DiagnosticsNode;
...
 typedef TickerCallback = void Function(Duration elapsed);
 
 /// An interface implemented by classes that can vend [Ticker] objects.
 ///
+/// To obtain a [TickerProvider], consider mixing in either
+/// [TickerProviderStateMixin] (which always works)
+/// or [SingleTickerProviderStateMixin] (which is more efficient when it works)
+/// to make a [State] subclass implement [TickerProvider].
+/// You might then pass that [State] to lower-level widgets
+/// or other related objects.
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/scheduler/ticker.dart:29]: Might as well do some rewording here to follow the [style guide](https://github.com/flutter/flutter/blob/58068d8ea4eca8a17d99469697aaff46255522e9/docs/contributing/Style-guide-for-Flutter-repo.md#use-the-passive-voice-recommend-do-not-require-never-say-things-are-simple).

```suggestion
/// The [State], or its constituent [AnimationController]s,
/// might then be passed to lower-level widgets or other related objects.
```

File: packages/flutter/lib/src/scheduler/ticker.dart [hunk 11]
```diff
@@ -43,9 +48,9 @@ abstract class TickerProvider {
   Ticker createTicker(TickerCallback onTick);
 }
 
-// TODO(jacobr): make Ticker use Diagnosticable to simplify reporting errors
-// related to a ticker.
-/// Calls its callback once per animation frame.
+/// Calls its callback once per animation frame, when enabled.
+///
+/// To obtain a ticker, consider [TickerProvider].
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/scheduler/ticker.dart:53]:
```suggestion
/// Calls its [TickerCallback] once per animation frame, when enabled.
```

The `TickerProvider` reference could be moved to a "See also" section at the bottom, for consistency with other docs.

File: packages/flutter/lib/src/scheduler/ticker.dart [hunk 12]
```diff
@@ -55,11 +60,15 @@ abstract class TickerProvider {
...
 /// By convention, the [start] and [stop] methods are used by the ticker's
-/// consumer, and the [muted] property is controlled by the [TickerProvider]
-/// that created the ticker.
+/// consumer (for example, an [AnimationController]), and the [muted] property
+/// is controlled by the [TickerProvider] that created the ticker (for example,
+/// a [State] that uses [TickerProviderStateMixin] to silence the ticker when
+/// the state's subtree is disabled as defined by [TickerMode]).
 ///
 /// Tickers are driven by the [SchedulerBinding]. See
 /// [SchedulerBinding.scheduleFrameCallback].
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/scheduler/ticker.dart:69]:
```suggestion
/// See also:
///
/// * [TickerProvider], which has instructions for obtaining a ticker.
/// * [SchedulerBinding] and its [SchedulerBinding.scheduleFrameCallback]
///   method, used to drive the ticker.
```

**Review by nate-thegrate — COMMENTED** [review]
This looks fantastic overall; thanks for putting the work in here.

File: packages/flutter/lib/src/animation/animation.dart (hunk 4 above)
Comment (gnprice) [reply packages/flutter/lib/src/animation/animation.dart:119]: Sure, good idea.

I expanded the heading slightly to "Performance considerations", following other examples in the codebase.

File: packages/flutter/lib/src/animation/animation.dart (hunk 7 above)
Comment (gnprice) [reply packages/flutter/lib/src/animation/animation.dart:125]: I don't feel strongly on this line but I think the "even while" version is slightly clearer. Otherwise it reads equally well as if it's saying that this optimization avoids rebuilding those widgets on each frame while the animation is active, but when it's inactive *they still get rebuilt*. (As opposed to the intended reading, where when it's inactive *it wouldn't be causing rebuilds anyway*.)

That other reading doesn't ultimately make sense in context. But the reader may be confused for a bit because the docs seem to be saying something that doesn't make sense, so that the reader has to sort out if the error is in their understanding of how things work, or in the docs, or (as it would turn out in this case) that they've hit an unintended reading of this sentence of the docs and there's a different reading that does make sense.

File: packages/flutter/lib/src/animation/animation_controller.dart (hunk 8 above)
Comment (gnprice) [reply packages/flutter/lib/src/animation/animation_controller.dart]: Yeah. With this version that information is one click away ("For advice … see [TickerProvider]."). I'm hoping to avoid introducing here both of those rather long names, plus the details of how to decide which one to use.

Tweaked to mention as an example SingleTickerProviderStateMixin, which I think is what one most often wants:
```dart
/// For advice on obtaining a ticker provider, see [TickerProvider].
/// Typically the relevant [State] serves as the ticker provider,
/// after applying a suitable mixin (like [SingleTickerProviderStateMixin])
/// to cause the [State] subclass to implement [TickerProvider].
```

File: packages/flutter/lib/src/animation/tween.dart (hunk 9 above)
Comment (gnprice) [reply packages/flutter/lib/src/animation/tween.dart]: The intended meaning of "evaluate" here is function evaluation, which isn't what "based on" would mean to me. But this seems like a sign this can be made clearer. Here's a version tweaked and made more explicit:

```suggestion
  /// the given parent and then evaluating this object at the result.
  ///
  /// This method represents function composition on [transform]:
  /// the [transform] method of the returned [Animatable] is the result of
  /// composing this object's [transform] method with
  /// the given parent's [transform] method.
```

File: packages/flutter/lib/src/scheduler/ticker.dart (hunk 10 above)
Comment (gnprice) [reply packages/flutter/lib/src/scheduler/ticker.dart]: Thanks, yeah. Rewritten like this:
```suggestion
/// That [State] can then be passed to lower-level widgets
/// or other related objects.
```

The [AnimationController]s might also be passed down; but that wouldn't be a way of obtaining a [TickerProvider], which is what this paragraph is about.

File: packages/flutter/lib/src/scheduler/ticker.dart (hunk 11 above)
Comment (gnprice) [reply packages/flutter/lib/src/scheduler/ticker.dart:53]: I think mentioning TickerProvider up at the top is helpful here because when someone's just trying to get an animation working and finding their way around the API, that's the reader who can most benefit from an early signpost helping them redirect toward what they're looking for.

File: packages/flutter/lib/src/scheduler/ticker.dart (hunk 12 above)
Comment (gnprice) [reply packages/flutter/lib/src/scheduler/ticker.dart]: Sure, good idea. Made a version that's a bit shorter:

```dart
/// See also:
/// 
/// * [TickerProvider], for obtaining a ticker.
/// * [SchedulerBinding.scheduleFrameCallback], which drives tickers.
```

**Review by gnprice — COMMENTED** [review]
Thanks for the reviews! Revision pushed; PTAL.

File: packages/flutter/lib/src/animation/animation.dart (hunk 7 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/animation/animation.dart:127]: After re-reading, I agree—it's pretty cool to only rebuild leaf widgets while there's an active animation!

File: packages/flutter/lib/src/scheduler/ticker.dart (hunk 11 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/scheduler/ticker.dart:56]: Fair point!

I still do think it'd be nice to link to the typedef here.

```suggestion
/// Calls its [TickerCallback] once per animation frame, when enabled.
///
/// To obtain a ticker, consider using a [TickerProvider].
```

And this is very pedantic, but I also feel that a sentence flows better when adding an "-ing" verb after "consider" (other alternatives: "consider the" or "consider who/what/when/where/how").

**Review by nate-thegrate — APPROVED** [review]
LGTM with 1 small nit.

Thanks again for doing this!

File: packages/flutter/lib/src/scheduler/ticker.dart (hunk 11 above)
Comment (gnprice) [reply packages/flutter/lib/src/scheduler/ticker.dart:56]: For the summary line, I think the style I typically see in the Flutter docs tends to keep that summary line in plain English prose (no identifiers) except where an identifier reference is critical to understanding the meaning.

So for example [TickerCallback] itself says in its summary:

```dart
/// Signature for the callback passed to the [Ticker] class's constructor.
```

which fits because that type is really all about the `Ticker` class.

But here, I think the exact signature of the callback isn't critical to a summary-level understanding of what `Ticker` is about. So that detail is best left for later, past the summary. (In the [docs page](https://main-api.flutter.dev/flutter/scheduler/Ticker-class.html), it'll appear right after the class's doc comment, in the signature of the constructor.)

---

On "consider", I feel like I see the terser form a lot in the Flutter docs, and I think it works well when the object of it is short, like here.

Looking in the Git history confirms that that wording is often from Hixie (who wrote much of Flutter's documentation in general and set its style):
```
$ git log -p --author hixie -G 'consider\b'
    # or even, filtering to some of the most core docs:
$ git log -p --author hixie -G 'consider\b' lib/src/widgets/framework.dart
```

> consider the [AppLifecycleListener] API.

> consider a single [CustomPaint] widget.

> consider an expression such as the following: [… a one-line code block …]

> consider [SelectionArea].

and so on. Certainly plenty of "-ing" gerund forms too, but usually when there's more to say about it.

**Review by gnprice — COMMENTED** [review]
Thanks!
